### PR TITLE
[1.21] Add a system to register custom ReceivingLevelScreens

### PR DESCRIPTION
**Pull request by GizmoTheMoonPig**
Hello! In 1.20.5, Mojang changed how the `ReceivingLevelScreen` displays when traveling across dimensions. When going to the nether/end, the portal effect is rendered over your screen instead of just the dirt background. As someone who works on 4 different dimension mods, I noticed that there is not currently a good way to add your own effects to this screen. You can either use a mixin (which I try to avoid like the plague) or cancel rendering in one of the `ScreenEvent`s and replace it with your own, which is honestly not ideal. This PR hopes to solve this problem. 

This PR adds a new client event known as `RegisterDimensionTransitionEvent` which allows you to set up custom rendering when entering or leaving a certain dimension. You can use this event to change what renders on screen, whether it's the actual screen effect or the "Downloading Terrain" text that appears. This should hopefully add a friendlier way for modders to customize their dimension transition screens. 

I will note I'm awful at writing Javadocs and naming things, so if you want to see anything changed please let me know. I also tried to keep patching minimal but I do need to pass the level keys around so the system knows what dimensions the player is moving to and from.

### CODE REVIEW

File: patches/net/minecraft/client/gui/screens/ReceivingLevelScreen.java.patch [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+ 
+     @Override
+@@ -41,11 +_,15 @@
+     @Override
+     public void render(GuiGraphics p_281489_, int p_282902_, int p_283018_, float p_281251_) {
+         super.render(p_281489_, p_282902_, p_283018_, p_281251_);
+-        p_281489_.drawCenteredString(this.font, DOWNLOADING_TERRAIN_TEXT, this.width / 2, this.height / 2 - 50, 16777215);
++        var screen = net.neoforged.neoforge.client.DimensionTransitionManager.get(this.toDimension, this.fromDimension);
++        if (screen != null) screen.renderTransitionText(this.font, p_281489_, this.width, this.height);
++        else p_281489_.drawCenteredString(this.font, DOWNLOADING_TERRAIN_TEXT, this.width / 2, this.height / 2 - 50, 16777215);
```
Comment (HenryLoenwind) [inline patches/net/minecraft/client/gui/screens/ReceivingLevelScreen.java.patch]: if you put the "else" on its own line, this patch chunk becomes a pure addition.

File: src/main/java/net/neoforged/neoforge/client/DimensionTransitionManager.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+    @ApiStatus.Internal
+    static void init() {
+        Map<ResourceKey<Level>, DimensionTransitionScreen> populatedToEffects = new HashMap<>();
+        Map<ResourceKey<Level>, DimensionTransitionScreen> populatedFromEffects = new HashMap<>();
+        RegisterDimensionTransitionEvent event = new RegisterDimensionTransitionEvent(populatedToEffects, populatedFromEffects);
+        ModLoader.postEventWrapContainerInModOrder(event);
+        toDimensionTransitions = populatedToEffects;
+        fromDimensionTransitions = populatedFromEffects;
+    }
```
Comment (HenryLoenwind) [inline src/main/java/net/neoforged/neoforge/client/DimensionTransitionManager.java]: Why all this hubbub with the maps?

```suggestion
    private static final Map<ResourceKey<Level>, DimensionTransitionScreen> toDimensionTransitions = new HashMap<>();
    private static final Map<ResourceKey<Level>, DimensionTransitionScreen> fromDimensionTransitions = new HashMap<>();

    @ApiStatus.Internal
    static void init() {
        ModLoader.postEventWrapContainerInModOrder(new RegisterDimensionTransitionEvent(toDimensionTransitions, fromDimensionTransitions);
    }
```

would accomplish exactly the same.

File: patches/net/minecraft/client/Minecraft.java.patch [hunk 3]
```diff
@@ -326,14 +326,16 @@
...
      private boolean isLevelRunningNormally() {
-@@ -2056,6 +_,7 @@
+@@ -2056,7 +_,8 @@
      }
  
      public void setLevel(ClientLevel p_91157_, ReceivingLevelScreen.Reason p_341652_) {
+-        this.updateScreenAndTick(new ReceivingLevelScreen(() -> false, p_341652_));
 +        if (this.level != null) net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.level.LevelEvent.Unload(this.level));
-         this.updateScreenAndTick(new ReceivingLevelScreen(() -> false, p_341652_));
++        this.updateScreenAndTick(new ReceivingLevelScreen(() -> false, p_341652_, this.level == null ? null : p_91157_.dimension(), this.level == null ? null : this.level.dimension()));
```
Comment (HenryLoenwind) [inline patches/net/minecraft/client/Minecraft.java.patch]: Wouldn't it be simpler to register a `BiFunction<BooleanSupplier, ReceivingLevelScreen.Reason, ? extends ReceivingLevelScreen>` (and return `ReceivingLevelScreen::new` instead of null from DimensionTransitionManager hwen there's noting registered)? That class doesn't look like it would present any hurdles to subclassing.

File: src/main/java/net/neoforged/neoforge/client/DimensionTransitionManager.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+    @Nullable
+    public static DimensionTransitionScreen get(@Nullable ResourceKey<Level> toDimension, @Nullable ResourceKey<Level> fromDimension) {
+        if (toDimension != null && toDimensionTransitions.containsKey(toDimension)) {
+            return toDimensionTransitions.get(toDimension);
+        } else if (fromDimension != null && fromDimensionTransitions.containsKey(fromDimension)) {
+            return fromDimensionTransitions.get(fromDimension);
+        }
+        return null;
```
Comment (HenryLoenwind) [inline src/main/java/net/neoforged/neoforge/client/DimensionTransitionManager.java]:
```suggestion
        return toDimensionTransitions.getOrDefault(toDimension, fromDimensionTransitions.getOrDefault(fromDimension, null));
```

As long as you're sure you have a HashMap, `get(null)` is fine.

File: src/main/java/net/neoforged/neoforge/client/event/RegisterDimensionTransitionEvent.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+import org.jetbrains.annotations.ApiStatus;
+
+/**
+ * <p>Event for registering screen effects when transitioning across dimensions.</p>
+ *
+ * <p>This event is not {@linkplain ICancellableEvent cancellable}, and does not {@linkplain HasResult have a result}.
+ *
+ * <p>This event is fired on the mod-specific event bus, only on the {@linkplain LogicalSide#CLIENT logical client}.</p>
+ */
+public class RegisterDimensionTransitionEvent extends Event implements IModBusEvent {
```
Comment (sciwhiz12) [inline src/main/java/net/neoforged/neoforge/client/event/RegisterDimensionTransitionEvent.java]: This name might be confusing since there already exists a class in Minecraft called `DimensionTransition`. Since this class is used for registering `DimensionTransitionScreen`s, it should be named `RegisterDimensionTransitionScreenEvent`.

File: src/main/java/net/neoforged/neoforge/client/DimensionTransitionScreen.java [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+     * Called when vanilla renders the "Downloading Terrain" text on the {@link ReceivingLevelScreen}. <br>
+     * Use this to render custom text on the screen if you don't want to use vanilla's.
+     *
+     * @param font         the game font
+     * @param graphics     the gui graphics
+     * @param screenWidth  the width of the screen
+     * @param screenHeight the height of the screen
+     */
+    default void renderTransitionText(Font font, GuiGraphics graphics, int screenWidth, int screenHeight) {
+        graphics.drawCenteredString(font, Component.translatable("multiplayer.downloadingTerrain"), screenWidth / 2, screenHeight / 2 - 50, 16777215);
```
Comment (sciwhiz12) [inline src/main/java/net/neoforged/neoforge/client/DimensionTransitionScreen.java]: Replace the magic constant for the color with a hex constant.
```suggestion
        graphics.drawCenteredString(font, Component.translatable("multiplayer.downloadingTerrain"), screenWidth / 2, screenHeight / 2 - 50, 0xFFFFFF);
```

File: src/main/java/net/neoforged/neoforge/client/DimensionTransitionScreen.java [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+     * Render an effect on screen when transitioning to/from a dimension via the {@link ReceivingLevelScreen}. If this method returns true, the vanilla rendering will be
+     * canceled. Otherwise, the screen will render the menu panorama.
+     *
+     * @param graphics     the gui graphics
+     * @param partialTicks progress between ticks
+     * @param screenWidth  the width of the screen
+     * @param screenHeight the height of the screen
+     * @return true to cancel vanilla rendering
+     */
+    boolean renderScreenEffect(GuiGraphics graphics, float partialTicks, int screenWidth, int screenHeight);
```
Comment (sciwhiz12) [inline src/main/java/net/neoforged/neoforge/client/DimensionTransitionScreen.java]: The `partialTicks` parameter should be renamed to `partialTick` (note the singular), since the parameter denotes the progress between two ticks, i.e., the partial value of a tick.

File: src/main/java/net/neoforged/neoforge/client/DimensionTransitionScreen.java [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+
+    /**
+     * Called when vanilla renders the "Downloading Terrain" text on the {@link ReceivingLevelScreen}. <br>
+     * Use this to render custom text on the screen if you don't want to use vanilla's.
```
Comment (sciwhiz12) [inline src/main/java/net/neoforged/neoforge/client/DimensionTransitionScreen.java]: Line breaks (`<br>`) should not be used, either as substitute for proper paragraphs or otherwise. Since the second sentence doesn't need to stand alone, the line break should simply be removed. (If you wish, you can replace it with paragraph tags.)

File: tests/src/main/java/net/neoforged/neoforge/debug/client/DimensionTransitionTests.java [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+            event.registerTo(Level.END, new DimensionTransitionScreen() {
+                @Override
+                public boolean renderScreenEffect(GuiGraphics graphics, float partialTicks, int screenWidth, int screenHeight) {
+                    graphics.blit(END_BG, 0, 0, 0, 0.0F, 0.0F, screenWidth, screenHeight, 32, 32);
+                    return true;
+                }
+
+                @Override
+                public void renderTransitionText(Font font, GuiGraphics graphics, int screenWidth, int screenHeight) {
+                    graphics.drawString(font, "This displays when going to the end!", screenWidth / 2, screenHeight / 2 - 50, 16777215);
```
Comment (sciwhiz12) [inline tests/src/main/java/net/neoforged/neoforge/debug/client/DimensionTransitionTests.java]: Replace the magic constant `16777215` for the color with the equivalent hex constant `0xFFFFFF`. Also, use `drawCenteredString` here instead of `drawString`.

(Please apply the same changes to the nether transition screen.)

File: src/main/java/net/neoforged/neoforge/client/event/RegisterDimensionTransitionEvent.java [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+import net.minecraft.world.level.Level;
+import net.neoforged.bus.api.Event;
+import net.neoforged.bus.api.ICancellableEvent;
+import net.neoforged.fml.LogicalSide;
+import net.neoforged.fml.event.IModBusEvent;
+import net.neoforged.neoforge.client.DimensionTransitionScreen;
+import org.jetbrains.annotations.ApiStatus;
+
+/**
+ * <p>Event for registering screen effects when transitioning across dimensions.</p>
```
Comment (sciwhiz12) [inline src/main/java/net/neoforged/neoforge/client/event/RegisterDimensionTransitionEvent.java]: This should clarify which has priority: the dimension transition screen of the originating dimension, or that of the destination dimension?

File: src/main/java/net/neoforged/neoforge/client/event/RegisterDimensionTransitionEvent.java [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+    @ApiStatus.Internal
+    public RegisterDimensionTransitionEvent(Map<ResourceKey<Level>, DimensionTransitionScreen> toEffects, Map<ResourceKey<Level>, DimensionTransitionScreen> fromEffects) {
+        this.toEffects = toEffects;
+        this.fromEffects = fromEffects;
+    }
+
+    /**
+     * Registers a dimension transition when traveling to a dimension.
+     */
+    public void registerTo(ResourceKey<Level> dimension, DimensionTransitionScreen screen) {
```
Comment (sciwhiz12) [inline src/main/java/net/neoforged/neoforge/client/event/RegisterDimensionTransitionEvent.java]: Are the names clear enough? We could perhaps name these to `outgoing`/`incoming`. (Just a thought.)

File: tests/src/main/java/net/neoforged/neoforge/debug/client/DimensionTransitionTests.java [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+
+@ForEachTest(side = Dist.CLIENT, groups = DimensionTransitionTests.GROUP)
+public class DimensionTransitionTests {
+    public static final String GROUP = "dimension_transition";
+    public static final ResourceLocation NETHER_BG = ResourceLocation.withDefaultNamespace("textures/block/netherrack.png");
+    public static final ResourceLocation END_BG = ResourceLocation.withDefaultNamespace("textures/block/end_stone.png");
+
+    @EmptyTemplate
+    @TestHolder(description = "Tests if a custom dimension transition is properly displayed", enabledByDefault = true)
+    static void dimensionTransitions(DynamicTest test) {
```
Comment (sciwhiz12) [inline tests/src/main/java/net/neoforged/neoforge/debug/client/DimensionTransitionTests.java]: For completeness, this test should be split into two: one for the end dimension outgoing transition screen, and one for the nether dimension incoming transition screen. 

Currently, because the two are combined, confirming the success of one screen confirms the whole test, even if the other screen hasn't been tested yet.

File: src/main/java/net/neoforged/neoforge/client/DimensionTransitionManager.java
Comment (sciwhiz12) [inline src/main/java/net/neoforged/neoforge/client/DimensionTransitionManager.java:1]: Same comment as for the event -- this should be called `DimensionTransitionScreenManager`.

File: patches/net/minecraft/client/Minecraft.java.patch (hunk 3 above)
Comment (sciwhiz12) [reply patches/net/minecraft/client/Minecraft.java.patch]: @GizmoTheMoonPig, what do you think of this suggestion?

**Review by sciwhiz12 — CHANGES_REQUESTED** [review]

File: patches/net/minecraft/client/Minecraft.java.patch (hunk 3 above)
Comment (GizmoTheMoonPig) [reply patches/net/minecraft/client/Minecraft.java.patch]: Sorry I missed this. I was trying to decipher it last week but was unable to, then I got really busy on a mod port I'm working on. 
I'm afraid I don't quite understand what's being asked here. Might just be my exhausted brain but I can't comprehend what's being asked here at all. I know its stupid but if you could dumb it down a bit I would appreciate it

Comment (sciwhiz12) [reply patches/net/minecraft/client/Minecraft.java.patch]: I think the suggestion is to allow modders to instead subclass `ReceivingLevelScreen` and then register constructors for that screen (the `BiFunction` being a functional interface typed with the same arguments as `ReceivingLevelScreen`'s constructor, so `ReceivingLevelScreen::new`), or something along those lines.

Comment (GizmoTheMoonPig) [reply patches/net/minecraft/client/Minecraft.java.patch]: alright, I think I understand now. I went ahead and pushed a change to the system. Let me know if it looks alright!

**Review by sciwhiz12 — DISMISSED** [review]
Tests confirm it works. Will wait a day or two for @HenryLoenwind to comment back before merging.

File: src/main/java/net/neoforged/neoforge/client/DimensionTransitionScreenManager.java [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+public class DimensionTransitionScreenManager {
+    private static final Map<ResourceKey<Level>, BiFunction<BooleanSupplier, ReceivingLevelScreen.Reason, ReceivingLevelScreen>> toDimensionTransitions = new HashMap<>();
+    private static final Map<ResourceKey<Level>, BiFunction<BooleanSupplier, ReceivingLevelScreen.Reason, ReceivingLevelScreen>> fromDimensionTransitions = new HashMap<>();
+
+    @ApiStatus.Internal
+    static void init() {
+        ModLoader.postEventWrapContainerInModOrder(new RegisterDimensionTransitionScreenEvent(toDimensionTransitions, fromDimensionTransitions));
+    }
+
+    public static BiFunction<BooleanSupplier, ReceivingLevelScreen.Reason, ReceivingLevelScreen> getScreen(@Nullable ResourceKey<Level> toDimension, @Nullable ResourceKey<Level> fromDimension) {
```
Comment (XFactHD) [inline src/main/java/net/neoforged/neoforge/client/DimensionTransitionScreenManager.java]: I think this would be a lot cleaner with a custom functional interface (i.e. `ReceivingLevelScreenFactory`) instead of this insane generic

File: src/main/java/net/neoforged/neoforge/client/DimensionTransitionScreenManager.java [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+        return toDimensionTransitions.getOrDefault(toDimension, fromDimensionTransitions.getOrDefault(fromDimension, ReceivingLevelScreen::new));
```
Comment (XFactHD) [inline src/main/java/net/neoforged/neoforge/client/DimensionTransitionScreenManager.java]: I would prefer a structure similar to this for readability:
```java
var toDim = toDimensionTransitions.getOrDefault(toDimension);
if (toDim != null) {
    return toDim;
}
var fromDim = fromDimensionTransitions.getOrDefault(fromDimension);
if (fromDim != null) {
    return fromDim;
}
return null;
```

File: src/main/java/net/neoforged/neoforge/client/event/RegisterDimensionTransitionScreenEvent.java [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+/**
+ * <p>Event for registering screen effects when transitioning across dimensions.
+ * Note that there is a priority order when it comes to what screens are displayed.
+ * If a dimension has a screen that displays when entering it, that will have priority over a dimension that has one when you leave it. </p>
+ *
+ * <p>This event is not {@linkplain ICancellableEvent cancellable}, and does not {@linkplain HasResult have a result}.
+ *
+ * <p>This event is fired on the mod-specific event bus, only on the {@linkplain LogicalSide#CLIENT logical client}.</p>
+ */
+public class RegisterDimensionTransitionScreenEvent extends Event implements IModBusEvent {
```
Comment (XFactHD) [inline src/main/java/net/neoforged/neoforge/client/event/RegisterDimensionTransitionScreenEvent.java:29]: Would it be useful to allow registering a transition that only applies when both origin and destination match? It would use a `Pair<RK<Level>, RK<Level>>` as the key and would be checked before the "from effects" and "to effects".

**Review by XFactHD — CHANGES_REQUESTED** [review]

File: src/main/java/net/neoforged/neoforge/client/event/RegisterDimensionTransitionScreenEvent.java (hunk 15 above)
Comment (GizmoTheMoonPig) [reply src/main/java/net/neoforged/neoforge/client/event/RegisterDimensionTransitionScreenEvent.java:29]: I could do that, not sure if anyone would use something like that but it may be nice to have

File: patches/net/minecraft/client/multiplayer/ClientPacketListener.java.patch [hunk 16]
```diff
@@ -54,6 +70,18 @@
  
              if (p_337415_ instanceof CommandBlockEntity && this.minecraft.screen instanceof CommandBlockEditScreen) {
                  ((CommandBlockEditScreen)this.minecraft.screen).updateGui();
+@@ -1420,9 +_,9 @@
+         }
+     }
+ 
+-    private void startWaitingForNewLevel(LocalPlayer p_304688_, ClientLevel p_304528_, ReceivingLevelScreen.Reason p_341690_) {
++    private void startWaitingForNewLevel(LocalPlayer p_304688_, ClientLevel p_304528_, ReceivingLevelScreen.Reason p_341690_, @Nullable ResourceKey<Level> toDimension, @Nullable ResourceKey<Level> fromDimension) {
```
Comment (XFactHD) [inline patches/net/minecraft/client/multiplayer/ClientPacketListener.java.patch:83]: Vanilla methods should never be deleted, even if they are private. Therefore the vanilla method should remain as deprecated with a comment (i.e. `/** @deprecated Neo: use {@link #startWaitingForNewLevel(<insert param types here>)} instead */`) and forward to the Neo-added one with default values for the added params (i.e. null in this case)

File: src/main/java/net/neoforged/neoforge/client/event/RegisterDimensionTransitionScreenEvent.java [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+    private final Map<Pair<ResourceKey<Level>, ResourceKey<Level>>, DimensionTransitionScreenManager.ReceivingLevelScreenFactory> conditionalDimensionEffects;
+    private final Map<ResourceKey<Level>, DimensionTransitionScreenManager.ReceivingLevelScreenFactory> toEffects;
+    private final Map<ResourceKey<Level>, DimensionTransitionScreenManager.ReceivingLevelScreenFactory> fromEffects;
+
+    @ApiStatus.Internal
+    public RegisterDimensionTransitionScreenEvent(Map<Pair<ResourceKey<Level>, ResourceKey<Level>>, DimensionTransitionScreenManager.ReceivingLevelScreenFactory> conditionalDimensionEffects, Map<ResourceKey<Level>, DimensionTransitionScreenManager.ReceivingLevelScreenFactory> toEffects, Map<ResourceKey<Level>, DimensionTransitionScreenManager.ReceivingLevelScreenFactory> fromEffects) {
```
Comment (XFactHD) [inline src/main/java/net/neoforged/neoforge/client/event/RegisterDimensionTransitionScreenEvent.java]: Please split the parameter list into multiple lines (make sure Spotless likes the formatting...) and consider importing the inner class separately.

**Review by XFactHD — CHANGES_REQUESTED** [review]

**Review by XFactHD — DISMISSED** [review]
Looks good to me. I'll defer back to sci's comment regarding HenryLoenwind's review

File: patches/net/minecraft/client/Minecraft.java.patch [hunk 18]
```diff
@@ -326,14 +326,16 @@
...
      private boolean isLevelRunningNormally() {
-@@ -2056,6 +_,7 @@
+@@ -2056,7 +_,8 @@
      }
  
      public void setLevel(ClientLevel p_91157_, ReceivingLevelScreen.Reason p_341652_) {
+-        this.updateScreenAndTick(new ReceivingLevelScreen(() -> false, p_341652_));
 +        if (this.level != null) net.neoforged.neoforge.common.NeoForge.EVENT_BUS.post(new net.neoforged.neoforge.event.level.LevelEvent.Unload(this.level));
-         this.updateScreenAndTick(new ReceivingLevelScreen(() -> false, p_341652_));
++        this.updateScreenAndTick(net.neoforged.neoforge.client.DimensionTransitionScreenManager.getScreen(this.level == null ? null : p_91157_.dimension(), this.level == null ? null : this.level.dimension()).create(() -> false, p_341652_));
```
Comment (HenryLoenwind) [inline patches/net/minecraft/client/Minecraft.java.patch]: `this.level == null ? null : p_91157_.dimension()` looks wrong to me. "If the current dimension is null, don't pass the new dimension"? If that's the intended logic, I suggest either adding a comment or moving the "null out target dimension" to the receiving method---here it looks like a "don't call a method on a null value" check (which I strongly suspect it is).

File: src/main/java/net/neoforged/neoforge/client/event/RegisterDimensionTransitionScreenEvent.java [hunk 19]
```diff
@@ -0,0 +1,72 @@
...
+        this.conditionalDimensionEffects = conditionalDimensionEffects;
+        this.toEffects = toEffects;
+        this.fromEffects = fromEffects;
+    }
+
+    /**
+     * Registers a dimension transition when traveling to a dimension.
+     */
+    public void registerIncomingEffect(ResourceKey<Level> dimension, ReceivingLevelScreenFactory screen) {
+        this.toEffects.put(dimension, screen);
```
Comment (HenryLoenwind) [inline src/main/java/net/neoforged/neoforge/client/event/RegisterDimensionTransitionScreenEvent.java]: I would suggest using putIfAbsent() and returning the result. Later mods silently overwriting earlier mods leads to hard to debug errors. With putIfAbsent(), the mod that gets its registration denied will at least get notified about it.

**_If_** there is a need to allow later mods to overwrite earlier ones, I'd add a third parameter `boolean overwrite` and `@returns if overwrite is false: true if the effect was registered, false if there already exists another effect for this transition; if overwrite is true: true if another registration was overwritten, false otherwise.`. A bit convoluted, so better find a nicer solution if needed...

Other opinions on this?

(PS: Same for the other register()s.)

**Review by HenryLoenwind — COMMENTED** [review]
Not sure about the second comment, I have a headache and am not in my best thinking form...

File: tests/src/main/java/net/neoforged/neoforge/debug/client/DimensionTransitionScreenTests.java [hunk 20]
```diff
@@ -0,0 +1,76 @@
...
+import net.minecraft.world.level.Level;
+import net.neoforged.api.distmarker.Dist;
+import net.neoforged.neoforge.client.event.RegisterDimensionTransitionScreenEvent;
+import net.neoforged.neoforge.event.entity.player.PlayerEvent;
+import net.neoforged.testframework.DynamicTest;
+import net.neoforged.testframework.annotation.ForEachTest;
+import net.neoforged.testframework.annotation.TestHolder;
+import net.neoforged.testframework.gametest.EmptyTemplate;
+
+@ForEachTest(side = Dist.CLIENT, groups = DimensionTransitionScreenTests.GROUP)
```
Comment (Matyrobbrt) [inline tests/src/main/java/net/neoforged/neoforge/debug/client/DimensionTransitionScreenTests.java]: Please put these tests in a `manual` group too (I want to introduce this pattern so that all manual tests are easier to run after a port). Additionally, is there a reason for the tests to be enabled by default? Unless there is one, it should be avoided as it will lead to test overload when you only want to test another test that's not related.

**Review by Matyrobbrt — CHANGES_REQUESTED** [review]

File: patches/net/minecraft/client/Minecraft.java.patch (hunk 18 above)
Comment (GizmoTheMoonPig) [reply patches/net/minecraft/client/Minecraft.java.patch]: It is intended as the level will be null on login. Vanilla only plays the transition screens when actually travelling to dimensions, not logging into them, so I replicated that behavior here. I will add a comment for clarity if desired

File: src/main/java/net/neoforged/neoforge/client/event/RegisterDimensionTransitionScreenEvent.java (hunk 19 above)
Comment (GizmoTheMoonPig) [reply src/main/java/net/neoforged/neoforge/client/event/RegisterDimensionTransitionScreenEvent.java]: I'd prefer if other mods didn't overwrite screens as this is more intended for things like dimension mods to add for their own stuff. I will switch to the putIfAbsent thing, but if someone has an actual use for overwriting screens I will accommodate for that

Comment (XFactHD) [reply src/main/java/net/neoforged/neoforge/client/event/RegisterDimensionTransitionScreenEvent.java]: I think returning a boolean indicating whether the registration was successful would be better, but other than that I agree with this.

File: patches/net/minecraft/client/Minecraft.java.patch (hunk 18 above)
Comment (HenryLoenwind) [reply patches/net/minecraft/client/Minecraft.java.patch]: Yes, this needs a comment.

How about you change getScreen() to take two Nullable Levels and put the logic there? There you have all the space you want to write a comment, here in the patch isn't really the place for comments.

File: src/main/java/net/neoforged/neoforge/client/event/RegisterDimensionTransitionScreenEvent.java [hunk 21]
```diff
@@ -0,0 +1,68 @@
...
+        this.fromEffects = fromEffects;
+    }
+
+    /**
+     * Registers a dimension transition when traveling to a dimension.
+     */
+    @Nullable
+    public ReceivingLevelScreenFactory registerIncomingEffect(ResourceKey<Level> dimension, ReceivingLevelScreenFactory screen) {
+        return this.toEffects.putIfAbsent(dimension, screen);
+    }
```
Comment (HenryLoenwind) [inline src/main/java/net/neoforged/neoforge/client/event/RegisterDimensionTransitionScreenEvent.java]:
```suggestion
    public boolean registerIncomingEffect(ResourceKey<Level> dimension, ReceivingLevelScreenFactory screen) {
        return this.toEffects.putIfAbsent(dimension, screen) == null;
    }
```

I should have been more specific. The caller only needs the information "success or failure", not the existing value.

File: src/main/java/net/neoforged/neoforge/client/event/RegisterDimensionTransitionScreenEvent.java [hunk 22]
```diff
@@ -0,0 +1,68 @@
...
+
+    /**
+     * Registers a dimension transition when traveling from a dimension.
+     */
+    @Nullable
+    public ReceivingLevelScreenFactory registerOutgoingEffect(ResourceKey<Level> dimension, ReceivingLevelScreenFactory screen) {
+        return this.fromEffects.putIfAbsent(dimension, screen);
+    }
```
Comment (HenryLoenwind) [inline src/main/java/net/neoforged/neoforge/client/event/RegisterDimensionTransitionScreenEvent.java]:
```suggestion
    /**
     * Registers a dimension transition when traveling from a dimension.
     *
     * @returns <code>true</code> if the screen was registered, <code>false</code> otherwise.
     */
    public boolean registerOutgoingEffect(ResourceKey<Level> dimension, ReceivingLevelScreenFactory screen) {
        return this.fromEffects.putIfAbsent(dimension, screen) == null;
    }
```

File: src/main/java/net/neoforged/neoforge/client/event/RegisterDimensionTransitionScreenEvent.java [hunk 23]
```diff
@@ -0,0 +1,68 @@
...
+
+    /**
+     * Registers a dimension transition when traveling to a dimension from a certain dimension.
+     * This registration method takes priority over the normal to and from dimension checks.
+     */
+    @Nullable
+    public ReceivingLevelScreenFactory registerConditionalEffect(ResourceKey<Level> toDimension, ResourceKey<Level> fromDimension, ReceivingLevelScreenFactory screen) {
+        return this.conditionalDimensionEffects.putIfAbsent(Pair.of(toDimension, fromDimension), screen);
+    }
```
Comment (HenryLoenwind) [inline src/main/java/net/neoforged/neoforge/client/event/RegisterDimensionTransitionScreenEvent.java]:
```suggestion
    /**
     * Registers a dimension transition when traveling to a dimension from a certain dimension.
     * This registration method takes priority over the normal to and from dimension checks.
     *
     * @returns <code>true</code> if the screen was registered, <code>false</code> otherwise.
     */
    public boolean registerConditionalEffect(ResourceKey<Level> toDimension, ResourceKey<Level> fromDimension, ReceivingLevelScreenFactory screen) {
        return this.conditionalDimensionEffects.putIfAbsent(Pair.of(toDimension, fromDimension), screen) == null;
    }
```

**Review by HenryLoenwind — CHANGES_REQUESTED** [review]

File: src/main/java/net/neoforged/neoforge/client/event/RegisterDimensionTransitionScreenEvent.java (hunk 22 above)
Comment (sciwhiz12) [reply src/main/java/net/neoforged/neoforge/client/event/RegisterDimensionTransitionScreenEvent.java]: BTW, use `{@code true}` rather than `<code>true</code>` in javadocs.

File: patches/net/minecraft/client/Minecraft.java.patch (hunk 18 above)
Comment (GizmoTheMoonPig) [reply patches/net/minecraft/client/Minecraft.java.patch]: The issue with changing it to use levels comes with the `ClientPacketListener` patch. `CommonPlayerSpawnInfo` doesn't provide a level, only a `ResourceKey`. That patch also checks that the player isn't dying so the transition screen doesn't display on respawn. I can still add the necessary comments in the `DimensionTransitionScreenManager` class, but I don't know if that will be as effective as adding the comment to the patch

Comment (HenryLoenwind) [reply patches/net/minecraft/client/Minecraft.java.patch]: You can always have two getScreen() methods. One that takes levels and goes

```
if source is null
    // initial load case, we don't want a transition screen for the target dim there
    return getSource(null, null)
else if target level is null
    // weird case, but better than an NPE, I guess...
    return getScreen(source.dim, null)
else
    return getScreen(source.dim, target.dim)
```

and the existing one.

**Review by HenryLoenwind — APPROVED** [review]

**Review by XFactHD — APPROVED** [review]

**Review by sciwhiz12 — APPROVED** [review]

**Review by Matyrobbrt — APPROVED** [review]
